src/synth.rs

```diff
@@ -0,0 +1,422 @@
+            .with_iter_limit(2)
+            .with_time_limit(Duration::from_secs(10))
+            .with_egraph(egraph)
+    }
+
+    fn run_rewrites(
+        &self,
+        mut runner: Runner<L, SynthAnalysis>,
+        rewrites: Vec<&Rewrite<L, SynthAnalysis>>,
+    ) -> (EGraph<L, SynthAnalysis>, HashMap<Id, Vec<Id>>, StopReason) {
+        let ids: Vec<Id> = runner.egraph.classes().map(|c| c.id).collect();
+        runner = runner.run(rewrites);
+        let stop_reason = runner.stop_reason.unwrap();
+
+        let mut found_unions = HashMap::default();
+        for id in ids {
+            let new_id = runner.egraph.find(id);
+            found_unions.entry(new_id).or_insert_with(Vec::new).push(id);
+        }
+
+        runner.egraph.rebuild();
+        (runner.egraph, found_unions, stop_reason)
+    }
+
+    fn apply_unions(&mut self, unions: HashMap<Id, Vec<Id>>) {
+        for ids in unions.values() {
+            if ids.len() > 1 {
+                let first = ids[0];
+                for id in &ids[1..] {
+                    self.egraph.union(first, *id);
+                }
+            }
+        }
+        self.egraph.rebuild();
+    }
+
+    fn cvec_match(&self) -> EqualityMap<L> {
+        // cvecs [𝑎1, . . . , 𝑎𝑛] and [𝑏1, . . . , 𝑏𝑛] match iff:
+        // ∀𝑖. 𝑎𝑖 = 𝑏𝑖 ∨ 𝑎𝑖 = null ∨ 𝑏𝑖 = null and ∃𝑖. 𝑎𝑖 = 𝑏𝑖 ∧ 𝑎𝑖 ≠ null ∧ 𝑏𝑖 ≠ null
+
+        println!(
+            "starting cvec match with {} eclasses",
+            self.egraph.number_of_classes()
+        );
+
+        let not_all_none: Vec<&EClass<L, Signature<L>>> = self
+            .egraph
+            .classes()
+            .filter(|x| x.data.cvec.iter().any(|v| v.is_some()))
+            .collect();
+
+        let compare = |cvec1: &CVec<L>, cvec2: &CVec<L>| -> bool {
+            for tup in cvec1.iter().zip(cvec2) {
+                match tup {
+                    (Some(a), Some(b)) if a != b => return false,
+                    _ => (),
+                }
+            }
+            true
+        };
+        let mut candidates = EqualityMap::default();
+        let extract = Extractor::new(&self.egraph, AstSize);
+        for class1 in &not_all_none {
+            for class2 in &not_all_none {
+                if class1.id == class2.id {
+                    continue;
+                }
+                if compare(&class1.data.cvec, &class2.data.cvec) {
+                    let (_, e1) = extract.find_best(class1.id);
+                    let (_, e2) = extract.find_best(class2.id);
+                    if let Some(eq) = Equality::new(&e1, &e2) {
+                        candidates.insert(eq.name.clone(), eq);
+                    }
+                    if let Some(eq) = Equality::new(&e2, &e1) {
+                        candidates.insert(eq.name.clone(), eq);
+                    }
+                }
+            }
+        }
+        candidates
+    }
+
+    // TODO: Figure out what to do with this- it doesn't match the definition
```

Comment:
IMHO this is fine. It definitely makes our tool incomplete but it is not totally unreasonable. Ruler is a synthesis tool, and such tools always have to work around state space explosion problems.